server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/VariableWidthHistogramAggregator.java

```diff
@@ -0,0 +1,378 @@
+        double clusterSize = clusterSizes.get(bucketOrd);
+        double newCentroid = ((oldCentroid * clusterSize) + val) / (clusterSize + 1);
+
+        clusterMaxes.set(bucketOrd, bucketMax);
+        clusterMins.set(bucketOrd, bucketMin);
+        clusterCentroids.set(bucketOrd, newCentroid);
+        clusterSizes.increment(bucketOrd, 1);
+    }
+
+    private void bucketCachedDocs() throws IOException{
+        Integer[] indexes = new Integer[(int)cachedValues.size()];
+        for(int i = 0; i < indexes.length; i++){
+            indexes[i] = i;
+        }
+
+        long[] mergeMap = new long[(int)cachedValues.size()];
+
+        // Sorting the documents by key enables bucketing with a single linear scan
+        // While sorting we have to track each document's initial index to determine which bucket its doc is in
+        // We can do this tracking by sorting the documents using their indexes
+        // After this step, this.cachedValues[this.indexes[i]] will be the i'th smallest value
+        Arrays.sort(indexes, Comparator.comparingDouble(idx -> cachedValues.get((long) idx)));
+
+        // An aggregator cannot return more than shardSize buckets.
+        // So leave room for some distant buckets to be added during rest of the collection phase.
+        int initialNumBuckets = shardSize * 3/4;
+
+        // Naively use basic linear separation to group the first cacheLimit docs into initialNumBuckets buckets
+        int docsPerBucket = (int) Math.ceil((double) cachedValues.size() / (double) initialNumBuckets);
+
+        int bucketOrd = 0;
+        for(int i = 0; i < indexes.length; i++){
+            // The doc that is currently in bucket indexes[i] is the i'th smallest
+            // We will move it into the the bucket at `bucketOrd`
+
+            int index = indexes[i];
+            double val = cachedValues.get(index);
+
+            mergeMap[index] = bucketOrd;
+            addToCluster(bucketOrd, val);
+
+            if((i + 1) % docsPerBucket == 0){
+                // There are enough docs in this bucket, so we should make a new one
+                bucketOrd += 1;
+            }
+        }
+
+        mergeBuckets(mergeMap, initialNumBuckets);
+        if (deferringCollector != null) {
+            deferringCollector.mergeBuckets(mergeMap);
+        }
+    }
+
+    /**
+     * Returns the ordinal of the bucket whose centroid is closest to <code>val</code>, or -1 if there are no buckets.
+     */
+    private int getNearestBucket(double val) {
+        double smallest_distance = Double.POSITIVE_INFINITY;
+        int nearest_bucket = -1;
+        for(int i = 0; i < clusterCentroids.size(); i++){
+            double current_distance = Math.abs(clusterCentroids.get(i) - val);
+            if(current_distance < smallest_distance){
+                smallest_distance = current_distance;
+                nearest_bucket = i;
+            }
+        }
+        return nearest_bucket;
```

Comment:
Would it make sense to keep the buckets in sorted order by centroid and binary search them? It'd certainly make inserting into the array more of a pain but that isn't something we do nearly as often as finding the nearest bucket.

Sad that we don't have a clean way to return the smallest distance. I think we'd want it at most of the call sites of this method.

Reply:
Good idea! I think that makes a lot of sense. I just implemented this fairly easily by using the existing `BucketsAggregator.mergeBuckets()` and  `MergingBucketsDeferringCollector.mergeBuckets()` methods 

If you like the overall change, I can make this more efficient by eliminating the need to create a merge map every time we want to insert. This would require creating very similar methods to the `mergeBuckets` ones, but that are specifically made for moving one bucket around. 

I didn't want spend time making this implementation more efficient before getting feedback on the idea. However, if you like it I can definitely so!

Reply:
I do like the idea. I'm starting to do work on `auto_date_histogram` lately and the whole `mergeMap` thing makes me sad in general. I know why we have it but it feels so wasteful! I'll read your changes and leave another comment when I'm through.